src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs

```diff
@@ -498,38 +499,41 @@ public void TexlFunctionTypeSemanticsIf()
                 symbol);
         }
 
-        [Fact]
-        public void TexlFunctionTypeSemanticsIfWithArgumentCoercion()
-        {
-            var symbol = new SymbolTable();
-            symbol.AddVariable("A", FormulaType.Number);
-            
-            TestSimpleBindingSuccess(
-                "If(A < 10, 1, \"2\")",
-                DType.Number,
-                symbol);
-
-            TestSimpleBindingSuccess(
-                "If(A < 1, \"one\", A < 2, 2, A < 3, true, false)",
-                DType.String,
-                symbol);
+        [Theory]
+        [InlineData("If(A < 10, 1, \"2\")", "n", true)]
+        [InlineData("If(A < 1, \"one\", A < 2, 2, A < 3, true, false)", "s", true)]
+        [InlineData("If(A < 1, true, A < 2, 2, A < 3, false, \"true\")", "b", true)]
+        [InlineData("If(A < 10, 1, [1,2,3])", "-", true)]
+        [InlineData("If(A < 10, 1, {Value: 2})", "-", true)]
+        [InlineData("If(0 < 1, [1], 2)", "-", true)]
 
-            TestSimpleBindingSuccess(
-                "If(A < 1, true, A < 2, 2, A < 3, false, \"true\")",
-                DType.Boolean,
-                symbol);
+        // negative cases, when if produces void type
 
-            // Negative cases -- when args cannot be coerced.
+        // void type is not allowed in aggregate type.
+        [InlineData("{test: If(A < 10, 1, {Value: 2})}", "![]", false)]
+        [InlineData("[If(A < 10, 1, {Value: 2})]", "*[]", false)]
 
-            TestBindingErrors(
-                "If(A < 10, 1, [1,2,3])",
-                DType.Number,
-                symbol);
+        // void type can't be consumed.
+        [InlineData("If(0 < 1, [1], 2) + 1", "n", false)]
+        public void TexlFunctionTypeSemanticsIfWithArgumentCoercion(string expression, string expectedType, bool checkSuccess)
+        {
+            var symbol = new SymbolTable();
+            symbol.AddVariable("A", FormulaType.Number);
 
-            TestBindingErrors(
-                "If(A < 10, 1, {Value: 2})",
-                DType.Number,
-                symbol);
+            if (checkSuccess)
+            {
+                TestSimpleBindingSuccess(
+                                    expression,
+                                    TestUtils.DT(expectedType),
+                                    symbol);
+            }
+            else
+            {
+                TestBindingErrors(
+                    expression,
+                    TestUtils.DT(expectedType),
+                    symbol);
+            }
         }
 
```

Comment:
We should have more negative tests, validating that a Void expression V will fail in:

- Functions that take number: `Abs(V)`
- Functions that take string: `Len(V)`
- Functions that take boolean: `If(V,1,2)`
- Functions that take date/time: `Hour(V)`
    - In the tests above, bonus points if the actual runtime value of is of the expected type, such as `Abs( If(false,{a:1},1 )`
- Records: `{a:V}`
- Tables: `[V]`
- ForAll expressions: `ForAll([1,2,3], V)`